File: ViewModels/FreemiumManager.swift
import Foundation
import SwiftUI
import Combine

class FreemiumManager: ObservableObject {
    @Published var showingSubscription = false {
        didSet {
            print("🔥🔥🔥 FREEMIUM MANAGER: showingSubscription changé vers \(showingSubscription)")
            // Envoyer une notification pour synchroniser avec MainView
            NotificationCenter.default.post(name: .freemiumManagerChanged, object: nil)
        }
    }
    @Published var blockedCategoryAttempt: QuestionCategory?
    
    private weak var appState: AppState?
    private var cancellables = Set<AnyCancellable>()
    
    // NOUVEAU: Configuration freemium
    private let questionsPerPack = 32
    private let freePacksLimit = 2 // 2 packs gratuits = 64 questions
    
    // NOUVEAU: Configuration freemium pour le journal
    private let freeJournalEntriesLimit = 5
    
    init(appState: AppState) {
        self.appState = appState
        setupObservers()
    }
    
    private func setupObservers() {
        // Observer les changements d'abonnement
        appState?.$currentUser
            .sink { [weak self] (user: AppUser?) in
                // Réagir aux changements d'abonnement si nécessaire
                print("🔥 FreemiumManager: Utilisateur changé - isSubscribed: \(user?.isSubscribed ?? false)")
            }
            .store(in: &cancellables)
    }
    
    // MARK: - Public Methods
    
    /// Vérifie si l'utilisateur peut accéder à une catégorie
    func canAccessCategory(_ category: QuestionCategory) -> Bool {
        // Les catégories premium nécessitent un abonnement
        if category.isPremium {
            return appState?.currentUser?.isSubscribed ?? false
        }
        
        // La catégorie "En couple" est gratuite
        return true
    }
    
    /// NOUVEAU: Vérifie si l'utilisateur peut accéder à une question spécifique dans une catégorie
    func canAccessQuestion(at index: Int, in category: QuestionCategory) -> Bool {
        // Si l'utilisateur est abonné, accès illimité
        if appState?.currentUser?.isSubscribed ?? false {
            return true
        }
        
        // Si c'est une catégorie premium, aucun accès
        if category.isPremium {
            return false
        }
        
        // Pour la catégorie "En couple" gratuite, limiter à 2 packs (64 questions)
        if category.id == "en-couple" {
            let maxFreeQuestions = freePacksLimit * questionsPerPack // 2 * 32 = 64
            return index < maxFreeQuestions
        }
        
        // Autres catégories gratuites (si elles existent)
        return true
    }
    
    /// NOUVEAU: Gère le tap sur une question avec vérification freemium
    func handleQuestionAccess(at index: Int, in category: QuestionCategory, onSuccess: @escaping () -> Void) {
        print("🔥🔥🔥 FREEMIUM QUESTION: Tentative accès question \(index + 1) dans \(category.title)")
        
        if canAccessQuestion(at: index, in: category) {
            print("🔥🔥🔥 FREEMIUM QUESTION: Accès autorisé")
            onSuccess()
        } else {
            print("🔥🔥🔥 FREEMIUM QUESTION: Accès bloqué - Affichage paywall")
            print("🔥🔥🔥 FREEMIUM QUESTION: Limite atteinte (64 questions) pour \(category.title)")
            
            blockedCategoryAttempt = category
            showingSubscription = true
            
            // Analytics - track blocked question
            trackQuestionBlocked(at: index, in: category)
        }
    }
    
    /// Gère le tap sur une catégorie avec la logique freemium
    func handleCategoryTap(_ category: QuestionCategory, onSuccess: @escaping () -> Void) {
        print("🔥 FreemiumManager: Tap sur catégorie: \(category.title)")
        print("🔥🔥🔥 FREEMIUM TAP: DEBUT GESTION TAP")
        print("🔥🔥🔥 FREEMIUM TAP: - Catégorie: \(category.title)")
        print("🔥🔥🔥 FREEMIUM TAP: - isPremium: \(category.isPremium)")
        print("🔥🔥🔥 FREEMIUM TAP: - Utilisateur abonné: \(appState?.currentUser?.isSubscribed ?? false)")
        print("🔥🔥🔥 FREEMIUM TAP: - showingSubscription AVANT: \(showingSubscription)")
        
        // Vérifier si l'utilisateur est abonné
        let isSubscribed = appState?.currentUser?.isSubscribed ?? false
        
        // Si l'utilisateur est abonné, accès illimité
        if isSubscribed {
            print("🔥🔥🔥 FREEMIUM TAP: UTILISATEUR ABONNE - ACCES ILLIMITE")
            onSuccess()
            return
        }
        
        // Si c'est une catégorie premium et l'utilisateur n'est pas abonné
        if category.isPremium {
            print("🔥🔥🔥 FREEMIUM TAP: CATEGORIE PREMIUM - ACCES BLOQUE")
            print("🔥 FreemiumManager: Accès bloqué à \(category.title) - affichage subscription")
            
            blockedCategoryAttempt = category
            print("🔥🔥🔥 FREEMIUM TAP: - blockedCategoryAttempt défini: \(category.title)")
            
            print("🔥🔥🔥 FREEMIUM TAP: - MISE A JOUR showingSubscription vers TRUE")
            showingSubscription = true
            print("🔥🔥🔥 FREEMIUM TAP: - showingSubscription APRES: \(showingSubscription)")
            
            // Notifier le changement
            NotificationCenter.default.post(name: .freemiumManagerChanged, object: nil)
            print("🔥🔥🔥 FREEMIUM TAP: - NOTIFICATION ENVOYEE")
            
            // Analytics - track blocked category
            trackCategoryBlocked(category)
            print("🔥🔥🔥 FREEMIUM TAP: FIN GESTION TAP BLOQUE")
            return
        }
        
        // Pour les catégories gratuites (comme "En couple"), permettre l'accès
        // La limitation se fera au niveau des questions dans QuestionListView
        print("🔥🔥🔥 FREEMIUM TAP: CATEGORIE GRATUITE - ACCES AUTORISE")
        print("🔥 FreemiumManager: Accès autorisé à \(category.title)")
        print("🔥🔥🔥 FREEMIUM TAP: ACCES AUTORISE - EXECUTION CALLBACK")
        onSuccess()
    }
    
    /// Retourne les catégories accessibles selon le statut d'abonnement
    func getAccessibleCategories() -> [QuestionCategory] {
        let isSubscribed = appState?.currentUser?.isSubscribed ?? false
        
        if isSubscribed {
            print("🔥 FreemiumManager: Utilisateur premium - toutes catégories accessibles")
            return QuestionCategory.categories
        } else {
            print("🔥 FreemiumManager: Utilisateur gratuit - catégories limitées")
            // Pour les utilisateurs gratuits, seule la première catégorie gratuite est accessible
            let freeCategories = QuestionCategory.categories.filter { !$0.isPremium }
            return Array(freeCategories.prefix(1)) // Seulement la première catégorie gratuite
        }
    }
    
    /// Retourne toutes les catégories avec leur statut de blocage
    func getAllCategoriesWithStatus() -> [QuestionCategory] {
        return QuestionCategory.categories
    }
    
    /// Vérifie si une catégorie doit être affichée comme bloquée
    func isCategoryBlocked(_ category: QuestionCategory) -> Bool {
        let isSubscribed = appState?.currentUser?.isSubscribed ?? false
        let isBlocked = category.isPremium && !isSubscribed
        
        print("🔥🔥🔥 FREEMIUM BLOCKED CHECK: Catégorie: \(category.title)")
        print("🔥🔥🔥 FREEMIUM BLOCKED CHECK: - isPremium: \(category.isPremium)")
        print("🔥🔥🔥 FREEMIUM BLOCKED CHECK: - isSubscribed: \(isSubscribed)")
        print("🔥🔥🔥 FREEMIUM BLOCKED CHECK: - isBlocked: \(isBlocked)")
        
        return isBlocked
    }
    
    /// NOUVEAU: Vérifie si une question doit être affichée comme bloquée
    func isQuestionBlocked(at index: Int, in category: QuestionCategory) -> Bool {
        return !canAccessQuestion(at: index, in: category)
    }
    
    /// NOUVEAU: Retourne le nombre maximum de questions gratuites pour une catégorie
    func getMaxFreeQuestions(for category: QuestionCategory) -> Int {
        if appState?.currentUser?.isSubscribed ?? false {
            return Int.max // Illimité pour les abonnés
        }
        
        if category.isPremium {
            return 0 // Aucune question gratuite pour les catégories premium
        }
        
        if category.id == "en-couple" {
            return freePacksLimit * questionsPerPack // 64 questions
        }
        
        return Int.max // Autres catégories gratuites (si elles existent)
    }
    
    /// Ferme la vue de subscription
    func dismissSubscription() {
        print("🔥 FreemiumManager: Fermeture de la vue subscription")
        print("🔥🔥🔥 FREEMIUM DISMISS: DEBUT FERMETURE")
        print("🔥🔥🔥 FREEMIUM DISMISS: - showingSubscription AVANT: \(showingSubscription)")
        print("🔥🔥🔥 FREEMIUM DISMISS: - blockedCategoryAttempt AVANT: \(blockedCategoryAttempt?.title ?? "nil")")
        
        showingSubscription = false
        blockedCategoryAttempt = nil
        
        // Notifier le changement
        NotificationCenter.default.post(name: .freemiumManagerChanged, object: nil)
        print("🔥🔥🔥 FREEMIUM DISMISS: - NOTIFICATION ENVOYEE")
        
        print("🔥🔥🔥 FREEMIUM DISMISS: - showingSubscription APRES: \(showingSubscription)")
        print("🔥🔥🔥 FREEMIUM DISMISS: - blockedCategoryAttempt APRES: \(blockedCategoryAttempt?.title ?? "nil")")
        print("🔥🔥🔥 FREEMIUM DISMISS: FIN FERMETURE")
    }
    
    /// NOUVEAU: Gère l'accès au widget de distance avec la logique freemium
    func handleDistanceWidgetAccess(onSuccess: @escaping () -> Void) {
        print("🔒 FreemiumManager: Accès au widget de distance demandé")
        
        // Tous les widgets sont maintenant gratuits
        print("✅ FreemiumManager: Accès autorisé au widget de distance (gratuit)")
        onSuccess()
    }
    
    /// NOUVEAU: Vérifie si l'utilisateur peut ajouter une nouvelle entrée journal
    func canAddJournalEntry(currentEntriesCount: Int) -> Bool {
        // Si l'utilisateur est abonné (direct ou hérité), pas de limite
        if appState?.currentUser?.isSubscribed ?? false {
            return true
        }
        
        // Pour les utilisateurs gratuits, limite à 5 entrées
        return currentEntriesCount < freeJournalEntriesLimit
    }
    
    /// NOUVEAU: Gère la tentative d'ajout d'entrée journal avec la logique freemium
    func handleJournalEntryCreation(currentEntriesCount: Int, onSuccess: @escaping () -> Void) {
        print("📝 FreemiumManager: Tentative d'ajout d'entrée journal")
        print("📝 FreemiumManager: Nombre d'entrées actuelles: \(currentEntriesCount)")
        
        // Vérifier si l'utilisateur est abonné (direct ou hérité)
        let isSubscribed = appState?.currentUser?.isSubscribed ?? false
        
        if isSubscribed {
            print("📝 FreemiumManager: Utilisateur premium - Ajout autorisé")
            onSuccess()
        } else if currentEntriesCount < freeJournalEntriesLimit {
            print("📝 FreemiumManager: Utilisateur gratuit - Ajout autorisé (\(currentEntriesCount)/\(freeJournalEntriesLimit))")
            onSuccess()
        } else {
            print("📝 FreemiumManager: Utilisateur gratuit - Limite atteinte (\(freeJournalEntriesLimit) entrées)")
            print("📝 FreemiumManager: Affichage du paywall pour journal")
            
            // Marquer que la tentative d'accès était pour le journal
            showingSubscription = true
            
            // Analytics - track blocked journal entry
            trackJournalEntryBlocked(entriesCount: currentEntriesCount)
        }
    }
    
    /// NOUVEAU: Retourne le nombre maximum d'entrées journal gratuites
    func getMaxFreeJournalEntries() -> Int {
        return freeJournalEntriesLimit
    }
    
    /// NOUVEAU: Retourne le nombre d'entrées restantes pour les utilisateurs gratuits
    func getRemainingFreeJournalEntries(currentEntriesCount: Int) -> Int {
        if appState?.currentUser?.isSubscribed ?? false {
            return Int.max // Illimité pour les abonnés
        }
        
        return max(0, freeJournalEntriesLimit - currentEntriesCount)
    }
    
    // MARK: - Analytics (pour le futur)
    
    private func trackCategoryBlocked(_ category: QuestionCategory) {
        print("🔥 FreemiumManager: Analytics - Catégorie bloquée: \(category.title)")
        // Ici vous pourrez ajouter Firebase Analytics, Mixpanel, etc.
        // Analytics.track("category_blocked", properties: ["category": category.title])
    }
    
    /// NOUVEAU: Analytics pour les questions bloquées
    private func trackQuestionBlocked(at index: Int, in category: QuestionCategory) {
        print("🔥 FreemiumManager: Analytics - Question bloquée: \(index + 1) dans \(category.title)")
        // Analytics.track("question_blocked", properties: ["category": category.title, "questionIndex": index])
    }
    
    /// NOUVEAU: Analytics pour le widget de distance bloqué
    private func trackDistanceWidgetBlocked() {
        print("🔒 FreemiumManager: Analytics - Widget de distance bloqué")
        // Analytics.track("distance_widget_blocked", properties: ["feature": "distance_widget"])
    }
    
    /// NOUVEAU: Analytics pour les entrées journal bloquées
    private func trackJournalEntryBlocked(entriesCount: Int) {
        print("📝 FreemiumManager: Analytics - Entrée journal bloquée à \(entriesCount) entrées")
        // Analytics.track("journal_entry_blocked", properties: ["entries_count": entriesCount, "limit": freeJournalEntriesLimit])
    }
    
    func trackUpgradePromptShown() {
        print("🔥 FreemiumManager: Analytics - Prompt d'upgrade affiché")
        // Analytics.track("upgrade_prompt_shown")
    }
    
    func trackConversion() {
        print("🔥 FreemiumManager: Analytics - Conversion réussie")
        // Analytics.track("freemium_conversion")
    }
}

// MARK: - Extensions pour le futur

extension FreemiumManager {
    /// Fonctionnalités futures - limitation quotidienne
    var dailyQuestionsRemaining: Int {
        // Pour l'instant, pas de limitation
        return 999
    }
    
    /// Fonctionnalités futures - système de crédits
    var credits: Int {
        // Pour l'instant, pas de système de crédits
        return 0
    }
    
    /// Fonctionnalités futures - accès temporaire
    func hasTemporaryAccess(to category: QuestionCategory) -> Bool {
        // Pour l'instant, pas d'accès temporaire
        return false
    }
} 

File: ViewModels/FreemiumManager.kt
package com.love2loveapp.core.viewmodels.freemium

import android.content.Context
import android.util.Log
import androidx.core.os.bundleOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.love2loveapp.core.common.AppException
import com.love2loveapp.core.common.Result
import com.love2loveapp.model.AppConstants
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

/**
 * Kotlin/Android version of the Swift FreemiumManager.
 *
 * - Observability: uses StateFlow/SharedFlow instead of @Published/NotificationCenter
 * - Analytics: FirebaseAnalytics
 * - Strings: Android strings.xml via context.getString(R.string.*)
 * - Time: uses epoch millis for persistence-friendly dates
 *
 * Expected strings.xml keys (create them if they don't exist):
 *  - daily_question_subtitle_subscribed
 *  - daily_question_subtitle_one_day_remaining
 *  - daily_question_subtitle_multiple_days_remaining  -> with "%1$d" placeholder
 *  - daily_question_subtitle_subscription_required
 *  - daily_challenge_subtitle_subscribed
 *  - daily_challenge_subtitle_one_day_remaining
 *  - daily_challenge_subtitle_multiple_days_remaining -> with "%1$d" placeholder
 *  - daily_challenge_subtitle_subscription_required
 */
class FreemiumManager(
    private val appState: AppState,
    private val context: Context
) : ViewModel() {

    private val TAG = "FreemiumManager"
    private val analytics: FirebaseAnalytics = FirebaseAnalytics.getInstance(context.applicationContext)

    // === Observable state ===
    private val _showingSubscription = MutableStateFlow(false)
    val showingSubscription: StateFlow<Boolean> = _showingSubscription.asStateFlow()

    private val _blockedCategoryAttempt = MutableStateFlow<QuestionCategory?>(null)
    val blockedCategoryAttempt: StateFlow<QuestionCategory?> = _blockedCategoryAttempt.asStateFlow()

    // Simple event bus to replace NotificationCenter.default.post(name: .freemiumManagerChanged, ...)
    private val _events = MutableSharedFlow<Event>(extraBufferCapacity = 1)
    val events: SharedFlow<Event> = _events.asSharedFlow()

    // === Freemium configuration (utilise constantes centralisées) ===
    private val questionsPerPack = AppConstants.Freemium.QUESTIONS_PER_PACK
    private val freePacksLimit = AppConstants.Freemium.FREE_PACKS_LIMIT
    private val freeJournalEntriesLimit = AppConstants.Freemium.FREE_JOURNAL_ENTRIES_LIMIT
    private val freeDailyQuestionDays = AppConstants.Freemium.FREE_DAILY_QUESTION_DAYS
    private val freeDailyChallengesDays = AppConstants.Freemium.FREE_DAILY_CHALLENGE_DAYS

    init {
        observeUserSubscriptionChanges()
    }

    private fun observeUserSubscriptionChanges() {
        viewModelScope.launch {
            appState.currentUserFlow.collect { user ->
                Log.d(TAG, "🔥 Utilisateur changé - isSubscribed: ${'$'}{user?.isSubscribed ?: false}")
            }
        }
    }

    private fun emitChanged() {
        _events.tryEmit(Event.FreemiumManagerChanged)
    }

    // === Public API (parité Swift) ===

    fun canAccessCategory(category: QuestionCategory): Boolean {
        // Premium categories require subscription
        if (category.isPremium) return appState.currentUser?.isSubscribed == true
        // "En couple" is free by design in your Swift code
        return true
    }

    fun canAccessQuestion(index: Int, category: QuestionCategory): Boolean {
        // Unlimited if subscribed
        if (appState.currentUser?.isSubscribed == true) return true

        // Block entire premium categories when not subscribed
        if (category.isPremium) return false

        // For the free "en-couple" category: limit to 2 packs * 32 = 64
        return if (category.id == "en-couple") {
            val maxFree = freePacksLimit * questionsPerPack
            index < maxFree
        } else {
            true // other free categories, if any
        }
    }

    fun handleQuestionAccess(index: Int, category: QuestionCategory, onSuccess: () -> Unit) {
        Log.d(TAG, "🔥 QUESTION: Tentative accès question ${'$'}{index + 1} dans ${'$'}{category.title}")
        if (canAccessQuestion(index, category)) {
            Log.d(TAG, "🔥 QUESTION: Accès autorisé")
            onSuccess()
        } else {
            Log.d(TAG, "🔥 QUESTION: Accès bloqué - Affichage paywall (limite atteinte)")
            _blockedCategoryAttempt.value = category
            setShowingSubscription(true)

            // Analytics
            analytics.logEvent("paywall_affiche", bundleOf(
                "source" to "freemium_limite"
            ))
            Log.d(TAG, "📊 FirebaseAnalytics: paywall_affiche — source=freemium_limite")

            trackQuestionBlocked(index, category)
        }
    }

    fun handleCategoryTap(category: QuestionCategory, onSuccess: () -> Unit) {
        val isSubscribed = appState.currentUser?.isSubscribed == true
        Log.d(TAG, "🔥 TAP: Catégorie='${'$'}{category.title}', isPremium=${'$'}{category.isPremium}, isSubscribed=${'$'}isSubscribed, showingSubscription=${'$'}{_showingSubscription.value}")

        if (isSubscribed) {
            Log.d(TAG, "🔥 TAP: UTILISATEUR ABONNÉ - ACCÈS ILLIMITÉ")
            onSuccess()
            return
        }

        if (category.isPremium) {
            Log.d(TAG, "🔥 TAP: CATEGORIE PREMIUM - ACCÈS BLOQUÉ → Paywall")
            _blockedCategoryAttempt.value = category
            setShowingSubscription(true)

            analytics.logEvent("paywall_affiche", bundleOf(
                "source" to "freemium_limite"
            ))
            Log.d(TAG, "📊 FirebaseAnalytics: paywall_affiche — source=freemium_limite")

            trackCategoryBlocked(category)
            return
        }

        // Free categories: allowed (per question limits enforced elsewhere)
        Log.d(TAG, "🔥 TAP: CATEGORIE GRATUITE - ACCÈS AUTORISÉ")
        onSuccess()
    }

    fun getAccessibleCategories(all: List<QuestionCategory>): List<QuestionCategory> {
        val isSubscribed = appState.currentUser?.isSubscribed == true
        return if (isSubscribed) {
            Log.d(TAG, "🔥 Utilisateur premium — toutes catégories accessibles")
            all
        } else {
            Log.d(TAG, "🔥 Utilisateur gratuit — catégories limitées (seulement la 1ère gratuite)")
            all.filter { !it.isPremium }.take(1)
        }
    }

    fun getAllCategoriesWithStatus(all: List<QuestionCategory>): List<QuestionCategory> = all

    fun isCategoryBlocked(category: QuestionCategory): Boolean {
        val isSubscribed = appState.currentUser?.isSubscribed == true
        val isBlocked = category.isPremium && !isSubscribed
        Log.d(TAG, "🔥 BLOCKED CHECK: Cat='${'$'}{category.title}', isPremium=${'$'}{category.isPremium}, isSubscribed=${'$'}isSubscribed, isBlocked=${'$'}isBlocked")
        return isBlocked
    }

    fun isQuestionBlocked(index: Int, category: QuestionCategory): Boolean = !canAccessQuestion(index, category)

    fun getMaxFreeQuestions(category: QuestionCategory): Int {
        if (appState.currentUser?.isSubscribed == true) return Int.MAX_VALUE
        if (category.isPremium) return 0
        return if (category.id == "en-couple") freePacksLimit * questionsPerPack else Int.MAX_VALUE
    }

    fun dismissSubscription() {
        Log.d(TAG, "🔥 DISMISS: Fermeture Subscription — before showing=${'$'}{_showingSubscription.value}, blocked='${'$'}{_blockedCategoryAttempt.value?.title ?: "nil"}'")
        _showingSubscription.value = false
        _blockedCategoryAttempt.value = null
        emitChanged()
        Log.d(TAG, "🔥 DISMISS: after showing=${'$'}{_showingSubscription.value}, blocked='${'$'}{_blockedCategoryAttempt.value?.title ?: "nil"}'")
    }

    fun handleDistanceWidgetAccess(onSuccess: () -> Unit) {
        Log.d(TAG, "🔒 Distance Widget: demandé → désormais gratuit")
        onSuccess()
    }

    fun canAddJournalEntry(currentEntriesCount: Int): Boolean {
        if (appState.currentUser?.isSubscribed == true) return true
        return currentEntriesCount < freeJournalEntriesLimit
    }

    fun handleJournalEntryCreation(currentEntriesCount: Int, onSuccess: () -> Unit) {
        Log.d(TAG, "📝 Journal: Tentative d'ajout — count=${'$'}currentEntriesCount")
        val isSubscribed = appState.currentUser?.isSubscribed == true
        when {
            isSubscribed -> {
                Log.d(TAG, "📝 Journal: Premium — Ajout autorisé")
                onSuccess()
            }
            currentEntriesCount < freeJournalEntriesLimit -> {
                Log.d(TAG, "📝 Journal: Gratuit — Ajout autorisé ${'$'}currentEntriesCount/${'$'}freeJournalEntriesLimit")
                onSuccess()
            }
            else -> {
                Log.d(TAG, "📝 Journal: Limite atteinte (${'$'}freeJournalEntriesLimit) → Paywall")
                setShowingSubscription(true)

                analytics.logEvent("paywall_affiche", bundleOf(
                    "source" to "popup_journal"
                ))
                Log.d(TAG, "📊 FirebaseAnalytics: paywall_affiche — source=popup_journal")

                trackJournalEntryBlocked(currentEntriesCount)
            }
        }
    }

    fun getMaxFreeJournalEntries(): Int = freeJournalEntriesLimit

    fun getRemainingFreeJournalEntries(currentEntriesCount: Int): Int {
        if (appState.currentUser?.isSubscribed == true) return Int.MAX_VALUE
        return (freeJournalEntriesLimit - currentEntriesCount).coerceAtLeast(0)
    }

    // === Daily Questions ===

    fun canAccessDailyQuestion(questionDay: Int): Boolean {
        if (appState.currentUser?.isSubscribed == true) return true
        Log.d(TAG, "📅 DailyQuestion: Vérification accès jour ${'$'}questionDay")
        return questionDay <= freeDailyQuestionDays
    }

    fun handleDailyQuestionAccess(currentQuestionDay: Int, onSuccess: () -> Unit) {
        val isSubscribed = appState.currentUser?.isSubscribed == true
        Log.d(TAG, "📅 DailyQuestion: Vérification jour ${'$'}currentQuestionDay, isSubscribed=${'$'}isSubscribed")
        if (isSubscribed) {
            markDailyQuestionUsage(currentQuestionDay)
            onSuccess()
            return
        }
        if (currentQuestionDay <= freeDailyQuestionDays) {
            Log.d(TAG, "📅 DailyQuestion: ${'$'}currentQuestionDay/${'$'}freeDailyQuestionDays — Accès gratuit")
            markDailyQuestionUsage(currentQuestionDay)
            onSuccess()
        } else {
            Log.d(TAG, "📅 DailyQuestion: Jour ${'$'}currentQuestionDay > ${'$'}freeDailyQuestionDays — Paywall")
            showDailyQuestionPaywall()
        }
    }

    private fun markDailyQuestionUsage(day: Int) {
        val user = appState.currentUser ?: return
        if (user.dailyQuestionFirstAccessDate == null) {
            user.dailyQuestionFirstAccessDate = System.currentTimeMillis()
            Log.d(TAG, "📅 DailyQuestion: Premier accès enregistré")
        }
        if (day > user.dailyQuestionMaxDayReached) {
            user.dailyQuestionMaxDayReached = day
            Log.d(TAG, "📅 DailyQuestion: Nouveau jour max atteint: ${'$'}day")
            appState.updateUser(user)
        }
    }

    private fun showDailyQuestionPaywall() {
        setShowingSubscription(true)
        analytics.logEvent("paywall_affiche", bundleOf(
            "source" to "daily_question_freemium"
        ))
        Log.d(TAG, "📊 FirebaseAnalytics: paywall_affiche — source=daily_question_freemium")
    }

    fun getDailyQuestionSubtitle(questionDay: Int): String {
        return if (appState.currentUser?.isSubscribed == true) {
            context.getString(R.string.daily_question_subtitle_subscribed)
        } else {
            if (questionDay <= freeDailyQuestionDays) {
                val remaining = (freeDailyQuestionDays - questionDay + 1).coerceAtLeast(0)
                if (remaining == 1) {
                    context.getString(R.string.daily_question_subtitle_one_day_remaining)
                } else {
                    context.getString(R.string.daily_question_subtitle_multiple_days_remaining, remaining)
                }
            } else {
                context.getString(R.string.daily_question_subtitle_subscription_required)
            }
        }
    }

    fun getRemainingFreeDaysForDailyQuestion(currentDay: Int): Int {
        if (appState.currentUser?.isSubscribed == true) return Int.MAX_VALUE
        return (freeDailyQuestionDays - currentDay + 1).coerceAtLeast(0)
    }

    // === Daily Challenges ===

    fun canAccessDailyChallenge(challengeDay: Int): Boolean {
        if (appState.currentUser?.isSubscribed == true) return true
        Log.d(TAG, "📅 DailyChallenge: Vérification accès défi jour ${'$'}challengeDay")
        return challengeDay <= freeDailyChallengesDays
    }

    fun handleDailyChallengeAccess(currentChallengeDay: Int, onSuccess: () -> Unit) {
        val isSubscribed = appState.currentUser?.isSubscribed == true
        Log.d(TAG, "📅 DailyChallenge: Vérification jour ${'$'}currentChallengeDay, isSubscribed=${'$'}isSubscribed")
        if (isSubscribed) {
            markDailyChallengeUsage(currentChallengeDay)
            onSuccess()
            return
        }
        if (currentChallengeDay <= freeDailyChallengesDays) {
            Log.d(TAG, "📅 DailyChallenge: ${'$'}currentChallengeDay/${'$'}freeDailyChallengesDays — Accès gratuit")
            markDailyChallengeUsage(currentChallengeDay)
            onSuccess()
        } else {
            Log.d(TAG, "📅 DailyChallenge: Jour ${'$'}currentChallengeDay > ${'$'}freeDailyChallengesDays — Paywall")
            showDailyChallengePaywall()
        }
    }

    private fun markDailyChallengeUsage(day: Int) {
        val user = appState.currentUser ?: return
        if (user.dailyChallengeFirstAccessDate == null) {
            user.dailyChallengeFirstAccessDate = System.currentTimeMillis()
            Log.d(TAG, "📅 DailyChallenge: Premier accès enregistré")
        }
        if (day > user.dailyChallengeMaxDayReached) {
            user.dailyChallengeMaxDayReached = day
            Log.d(TAG, "📅 DailyChallenge: Nouveau jour max atteint: ${'$'}day")
            appState.updateUser(user)
        }
        analytics.logEvent("freemium_daily_challenge_accessed", bundleOf(
            "challenge_day" to day,
            "is_subscribed" to false
        ))
    }

    private fun showDailyChallengePaywall() {
        setShowingSubscription(true)
        analytics.logEvent("paywall_viewed", bundleOf(
            "source" to "daily_challenge_freemium",
            "day" to 1
        ))
        analytics.logEvent("paywall_affiche", bundleOf(
            "source" to "daily_challenge_freemium"
        ))
        Log.d(TAG, "📊 FirebaseAnalytics: paywall_affiche — source=daily_challenge_freemium")
    }

    fun getDailyChallengeSubtitle(challengeDay: Int): String {
        return if (appState.currentUser?.isSubscribed == true) {
            context.getString(R.string.daily_challenge_subtitle_subscribed)
        } else {
            if (challengeDay <= freeDailyChallengesDays) {
                val remaining = (freeDailyChallengesDays - challengeDay + 1).coerceAtLeast(0)
                if (remaining == 1) {
                    context.getString(R.string.daily_challenge_subtitle_one_day_remaining)
                } else {
                    context.getString(R.string.daily_challenge_subtitle_multiple_days_remaining, remaining)
                }
            } else {
                context.getString(R.string.daily_challenge_subtitle_subscription_required)
            }
        }
    }

    fun getRemainingFreeDaysForDailyChallenge(currentDay: Int): Int {
        if (appState.currentUser?.isSubscribed == true) return Int.MAX_VALUE
        return (freeDailyChallengesDays - currentDay + 1).coerceAtLeast(0)
    }

    // === Analytics convenience ===

    private fun trackCategoryBlocked(category: QuestionCategory) {
        Log.d(TAG, "🔥 Analytics: Catégorie bloquée: ${'$'}{category.title}")
        // Add extra analytics here if needed
    }

    private fun trackQuestionBlocked(index: Int, category: QuestionCategory) {
        Log.d(TAG, "🔥 Analytics: Question bloquée: ${'$'}{index + 1} dans ${'$'}{category.title}")
        // Add extra analytics here if needed
    }

    private fun trackJournalEntryBlocked(entriesCount: Int) {
        Log.d(TAG, "📝 Analytics: Entrée journal bloquée à ${'$'}entriesCount entrées (limite=${'$'}freeJournalEntriesLimit)")
        // Add extra analytics here if needed
    }

    fun trackUpgradePromptShown() {
        Log.d(TAG, "🔥 Analytics: Prompt d'upgrade affiché")
        // analytics.logEvent("upgrade_prompt_shown", null)
    }

    fun trackConversion() {
        Log.d(TAG, "🔥 Analytics: Conversion réussie")
        // analytics.logEvent("freemium_conversion", null)
    }

    // === Helpers ===
    fun setShowingSubscription(value: Boolean) {
        _showingSubscription.value = value
        emitChanged()
        Log.d(TAG, "🔥 showingSubscription changé vers ${'$'}value")
    }

    fun setBlockedCategoryAttempt(category: QuestionCategory?) {
        _blockedCategoryAttempt.value = category
        emitChanged()
    }

    // === Result<T> based operations ===
    
    /**
     * Vérifie l'accès à une catégorie avec Result<T>
     */
    fun checkCategoryAccess(category: QuestionCategory): Result<Boolean> = try {
        val canAccess = canAccessCategory(category)
        if (canAccess) {
            Result.success(true)
        } else {
            Result.error(AppException.Data.ValidationError("premium_category_blocked"))
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }
    
    /**
     * Vérifie l'accès à une question avec Result<T>
     */
    fun checkQuestionAccess(index: Int, category: QuestionCategory): Result<Boolean> = try {
        val canAccess = canAccessQuestion(index, category)
        if (canAccess) {
            Result.success(true)
        } else {
            Result.error(AppException.Data.ValidationError("question_limit_reached"))
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }
    
    /**
     * Vérifie l'accès au journal avec Result<T>
     */
    fun checkJournalAccess(currentEntriesCount: Int): Result<Boolean> = try {
        val canAccess = canAddJournalEntry(currentEntriesCount)
        if (canAccess) {
            Result.success(true)
        } else {
            Result.error(AppException.Data.ValidationError("journal_limit_reached"))
        }
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }
    
    /**
     * État des catégories accessibles avec Result<T>
     */
    fun getAccessibleCategoriesResult(all: List<QuestionCategory>): Result<List<QuestionCategory>> = try {
        val accessible = getAccessibleCategories(all)
        Result.success(accessible)
    } catch (e: Exception) {
        Result.error(AppException.fromThrowable(e))
    }

    // === Event types ===
    sealed class Event {
        object FreemiumManagerChanged : Event()
    }
}

// === Minimal app-level contracts to mirror your Swift types ===

data class AppUser(
    var isSubscribed: Boolean = false,
    var dailyQuestionFirstAccessDate: Long? = null,
    var dailyQuestionMaxDayReached: Int = 0,
    var dailyChallengeFirstAccessDate: Long? = null,
    var dailyChallengeMaxDayReached: Int = 0,
)

interface AppState {
    val currentUserFlow: StateFlow<AppUser?>
    var currentUser: AppUser?
    fun updateUser(user: AppUser)
}

/**
 * Port of your QuestionCategory used in Swift.
 * Replace companion object with your real repository/provider.
 */
data class QuestionCategory(
    val id: String,
    val title: String,
    val isPremium: Boolean
) {
    companion object {
        // Replace with your actual categories source
        val categories: List<QuestionCategory> = emptyList()
    }
}
